src/main/java/crud/web/filters/AuthenticationFilter.java

```diff
@@ -0,0 +1,46 @@
+package crud.web.filters;
+
+import crud.lib.Injector;
+import crud.service.abstraction.DriverService;
+import java.io.IOException;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class AuthenticationFilter implements Filter {
+    private static final String DRIVER_ID = "driver_id";
+    private static final Injector INJECTOR = Injector.getInstance("crud");
+    private final DriverService driverService
+            = (DriverService) INJECTOR.getInstance(DriverService.class);
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest req = (HttpServletRequest) request;
+        HttpServletResponse resp = (HttpServletResponse) response;
+        String servletPath = req.getServletPath();
+        if (servletPath.equals("/login") || servletPath.equals("/drivers/create")) {
+            chain.doFilter(req, resp);
+            return;
+        }
+        Long driverId = (Long) req.getSession().getAttribute(DRIVER_ID);
+        if (driverId == null || driverService.get(driverId) == null) {
```

Comment:
does your service return null or it throws an exception?

Reply:
driver service returns null. I know, that returning null is a bad practice, but it was made to handle a situation where logged in user deletes themselves and for this user not to see an exception stacktrace in their browser.

Reply:
that makes program structure a little confusing (why do we need to return optional from dao to return null from service if there is not object). Let's not change the service layer and not handle this particular case. In real app you will use Spring security and it's possible to invalidate a users' session when you delete one so it won't be a problem.